adapters/metax/metax.go

```diff
@@ -0,0 +1,244 @@
+package metax
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"text/template"
+
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v2/adapters"
+	"github.com/prebid/prebid-server/v2/config"
+	"github.com/prebid/prebid-server/v2/errortypes"
+	"github.com/prebid/prebid-server/v2/macros"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+	"github.com/prebid/prebid-server/v2/util/ptrutil"
+)
+
+const SupportedCurrency = "USD"
+
+type adapter struct {
+	template *template.Template
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "No impressions in the request",
+		}}
+	}
+
+	// split impressions
+	reqDatas := make([]*adapters.RequestData, 0, len(request.Imp))
+	for i := range request.Imp {
+		imp := &request.Imp[i]
+		impCopy := *imp
+		requestCopy := *request
+
+		metaxExt, err := parseBidderExt(imp)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if err := validateParams(metaxExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if err := preprocessImp(&impCopy); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		endpoint, err := a.getEndpoint(metaxExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		impCopy.Ext = nil
+		requestCopy.Imp = []openrtb2.Imp{impCopy}
+		reqJSON, err := json.Marshal(requestCopy)
+		if err != nil {
+			errs = append(errs, err)
+			return nil, errs
+		}
+
+		headers := http.Header{}
+		headers.Add("Content-Type", "application/json;charset=utf-8")
+		headers.Add("Accept", "application/json")
+		reqDatas = append(reqDatas, &adapters.RequestData{
+			Method:  "POST",
+			Uri:     endpoint,
+			Body:    reqJSON,
+			Headers: headers,
+			ImpIDs:  openrtb_ext.GetImpIDs(requestCopy.Imp),
+		})
+	}
+
+	return reqDatas, errs
+}
+
+func (a *adapter) MakeBids(bidReq *openrtb2.BidRequest, reqData *adapters.RequestData, respData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if respData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if respData.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Bad request", respData.StatusCode),
+		}}
+	}
+
+	if respData.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d", respData.StatusCode),
+		}}
+	}
+
```

Comment:
could make use of response utils
https://github.com/prebid/prebid-server/blob/ec729e6e5d50a1306d76332fda52108b9d313c1d/adapters/response.go#L10-L28